ThunkLibs/libfex_thunk_test/Host.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+$info$
+tags: thunklibs|asound
+$end_info$
+*/
+
+#include <dlfcn.h>
+
+#include <chrono>
+#include <cstring>
+#include <new>
+#include <thread>
+#include <unordered_map>
+
+#include "common/Host.h"
+
+#include "api.h"
+
+#include "thunkgen_host_libfex_thunk_test.inl"
+
+#if 0
+std::thread* thr;
+
+void fexfn_impl_libfex_thunk_test_SetAsyncCallback(void(*cb)()) {
+  auto host_cb = (void (*)())cb;
+  FinalizeHostTrampolineForGuestFunction(host_cb);
+  MakeHostTrampolineForGuestFunctionAsyncCallable(host_cb, 1);
+
+  thr = new std::thread { [host_cb]() {
+    using namespace std::chrono_literals;
+    std::this_thread::sleep_for(500ms);
+    host_cb();
+  }};
+
+//  host_cb();
+  printf("HELLO WORLD\n");
+}
+#endif
+
+using const_void_ptr = const void*;
+
+template<StructType TypeIndex, typename Type>
+static const TestBaseStruct* convert(const TestBaseStruct* source) {
+    // Using malloc here since no easily available type information is available at the time of destruction
+    auto child = (host_layout<Type>*)malloc(sizeof(host_layout<Type>));
+    new (child) host_layout<Type> { *reinterpret_cast<guest_layout<Type>*>((void*)(source)) }; // TODO: Use proper cast?
+    // TODO: Trigger *full* custom repack for children, not just the Next member
+    child->data.Next = fex_custom_repack<&Type::Next>(child->data.Next);
+
+    return (const TestBaseStruct*)child; // TODO: Use proper cast?
+}
+
+template<StructType TypeIndex, typename Type>
+inline constexpr std::pair<StructType, const TestBaseStruct*(*)(const TestBaseStruct*)> converters =
+  { TypeIndex, convert<TypeIndex, Type> };
+
+
+static std::unordered_map<StructType, const TestBaseStruct*(*)(const TestBaseStruct*)> next_handlers {
+    converters<StructType::Struct1, TestStruct1>,
+    converters<StructType::Struct2, TestStruct2>,
+};
+
+// Normally, we would implement fex_custom_repack individually for each customized struct.
+// In this case, they all need the same repacking, so we just implement it once and alias all fex_custom_repack instances
+extern "C" const_void_ptr default_fex_custom_repack(const const_void_ptr& source) {
```

Comment:
I guess these shouldn't be references since checking for nullptr on a reference is...icky?

Reply:
Oop wait, it's a reference to a pointer type, nevermind.